autoprecompiles/src/memory_optimizer.rs

```diff
@@ -4,55 +4,59 @@ use std::hash::Hash;
 
 use itertools::Itertools;
 use powdr_constraint_solver::boolean_extractor::{self, RangeConstraintsForBooleans};
-use powdr_constraint_solver::constraint_system::{ConstraintRef, ConstraintSystem};
+use powdr_constraint_solver::constraint_system::{BusInteraction, ConstraintRef, ConstraintSystem};
 use powdr_constraint_solver::indexed_constraint_system::IndexedConstraintSystem;
-use powdr_constraint_solver::quadratic_symbolic_expression::RangeConstraintProvider;
 use powdr_constraint_solver::quadratic_symbolic_expression::{
-    NoRangeConstraints, QuadraticSymbolicExpression,
+    QuadraticSymbolicExpression, RangeConstraintProvider,
 };
 use powdr_constraint_solver::utils::possible_concrete_values;
-use powdr_number::{FieldElement, LargeInt};
+use powdr_number::FieldElement;
 
-use crate::legacy_expression::{AlgebraicExpression, AlgebraicReference};
-use crate::optimizer::algebraic_to_quadratic_symbolic_expression;
-use crate::{SymbolicBusInteraction, SymbolicConstraint, SymbolicMachine, MEMORY_BUS_ID};
+use crate::MEMORY_BUS_ID;
 
 /// The memory address space for register memory operations.
 const REGISTER_ADDRESS_SPACE: u32 = 1;
 
 /// Optimizes bus sends that correspond to general-purpose memory read and write operations.
 /// It works best if all read-write-operation addresses are fixed offsets relative to some
 /// symbolic base address. If stack and heap access operations are mixed, this is usually violated.
-pub fn optimize_memory<T: FieldElement>(mut machine: SymbolicMachine<T>) -> SymbolicMachine<T> {
-    let (to_remove, new_constraints) = redundant_memory_interactions_indices(&machine);
+pub fn optimize_memory<T: FieldElement, V: Hash + Eq + Clone + Ord + Display>(
```

Comment:
More of a general question: would is make sense to introduce a trait to group these bounds?

Reply:
I mean we don't always need all of those, but we can think about it.

Reply:
Maybe let's do that together with the SymbolicExpression abstraction